esp-hal-common/src/uart.rs

```diff
@@ -1079,16 +1098,51 @@ mod asynch {
     where
         T: Instance,
     {
+        pub async fn read(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+            let mut read_bytes = 0;
+            #[allow(unused_variables)]
+            let offset = 0;
+
+            // on ESP32-S2 we need to use PeriBus2 to read the FIFO
+            #[cfg(esp32s2)]
+            let offset = 0x20c00000;
+
+            UartFuture::new(Event::RxIntr, self.inner()).await;
+
+            while self.uart.get_rx_fifo_count() > 0 && read_bytes < buf.len() {
+                buf[read_bytes] = unsafe {
+                    let fifo = (self.uart.register_block().fifo.as_ptr() as *mut u8).offset(offset)
+                        as *mut crate::peripherals::generic::Reg<FIFO_SPEC>;
+                    (*fifo).read().rxfifo_rd_byte().bits()
+                };
+                read_bytes += 1;
+            }
+
+            Ok(read_bytes)
+        }
+
         async fn write(&mut self, words: &[u8]) -> Result<(), Error> {
             for chunk in words.chunks(UART_FIFO_SIZE as usize) {
                 for &byte in chunk {
-                    self.write_byte(byte).unwrap() // should never fail
```

Comment:
Though I can see a failure case with the original code, I don't think this change is correct. I think we should either:

- flush before writing
- read the number of bytes in the FIFO and ensure we only write the number of bytes left in the FIFO initially.